src/metabase/pulse.clj

```diff
@@ -106,174 +159,215 @@
-    (.setStroke g2 (new java.awt.BasicStroke sparkline-thickness java.awt.BasicStroke/CAP_ROUND java.awt.BasicStroke/JOIN_ROUND))
-    (.drawPolyline g2 (int-array (count xt) xt) (int-array (count yt) yt) (count xt))
-    (.setColor g2 (new java.awt.Color 45 134 212))
-    (.fillOval g2 (- (last xt) sparkline-dot-radius) (- (last yt) sparkline-dot-radius) (* 2 sparkline-dot-radius) (* 2 sparkline-dot-radius))
-    (.setColor g2 java.awt.Color/white)
-    (.setStroke g2 (new java.awt.BasicStroke 2))
-    (.drawOval g2 (- (last xt) sparkline-dot-radius) (- (last yt) sparkline-dot-radius) (* 2 sparkline-dot-radius) (* 2 sparkline-dot-radius))
-    (javax.imageio.ImageIO/write image "png" os)
+  (let [os    (ByteArrayOutputStream.)
+        image (BufferedImage. (+ width (* 2 sparkline-pad)) (+ height (* 2 sparkline-pad)) BufferedImage/TYPE_INT_ARGB)
+        xt    (map #(+ sparkline-pad (* width %)) xs)
+        yt    (map #(+ sparkline-pad (- height (* height %))) ys)]
+    (doto (.createGraphics image)
+      (.setRenderingHints (RenderingHints. RenderingHints/KEY_ANTIALIASING RenderingHints/VALUE_ANTIALIAS_ON))
+      (.setColor (Color. 211 227 241))
+      (.setStroke (BasicStroke. sparkline-thickness BasicStroke/CAP_ROUND BasicStroke/JOIN_ROUND))
+      (.drawPolyline (int-array (count xt) xt)
+                     (int-array (count yt) yt)
+                     (count xt))
+      (.setColor (Color. 45 134 212))
+      (.fillOval (- (last xt) sparkline-dot-radius)
+                 (- (last yt) sparkline-dot-radius)
+                 (* 2 sparkline-dot-radius)
+                 (* 2 sparkline-dot-radius))
+      (.setColor Color/white)
+      (.setStroke (BasicStroke. 2))
+      (.drawOval (- (last xt) sparkline-dot-radius)
+                 (- (last yt) sparkline-dot-radius)
+                 (* 2 sparkline-dot-radius)
+                 (* 2 sparkline-dot-radius)))
+    (ImageIO/write image "png" os)
     (.toByteArray os)))
 
-(defn render-card-sparkline
+(defn- render-card-sparkline
   [card {:keys [rows cols] :as data} render-img include-buttons]
-  (let [xs (for [row rows :let [x (first row)]] (if (instance? java.util.Date x) (.getTime x) x))
-        xmin (apply min xs)
-        xmax (apply max xs)
+  (let [xs     (for [row  rows
+                     :let [x (first row)]]
+                 (if (instance? Date x)
+                   (.getTime ^Date x)
+                   x))
+        xmin   (apply min xs)
+        xmax   (apply max xs)
         xrange (- xmax xmin)
-        xs' (map #(/ (- % xmin) xrange) xs)
-        ys (map second rows)
-        ymin (apply min ys)
-        ymax (apply max ys)
+        xs'    (map #(/ (- % xmin) xrange) xs)
+        ys     (map second rows)
+        ymin   (apply min ys)
+        ymax   (apply max ys)
         yrange (max 1 (- ymax ymin)) ; `(max 1 ...)` so we don't divide by zero
-        ys' (map #(/ (- % ymin) yrange) ys)
-        rows' (->> rows (take-last 2) (reverse))
```

Comment:
You don't need to put parens around single args to threading macros. You can do:
```clojure
(->> rows (take-last 2) reverse)
```

Also FWIW I don't think the threading macro makes this code any clearer than just writing
```clojure
(reverse (take-last 2 rows))
```